src/vopr/main.go

```diff
@@ -0,0 +1,782 @@
+		os.Exit(1)
+	} else if not_empty(&vopr_hub_address) {
+		log_debug("vopr_hub_address set as "+vopr_hub_address, nil)
+	} else {
+		log_error("VOPR_HUB_ADDRESS was empty", nil)
+		os.Exit(1)
+	}
+
+	repository_url, found = os.LookupEnv("REPOSITORY_URL")
+	if !found {
+		log_error("Could not find REPOSITORY_URL environmental variable", nil)
+		os.Exit(1)
+	} else if not_empty(&repository_url) {
+		log_debug("repository_url set as "+repository_url, nil)
+	} else {
+		log_error("REPOSITORY_URL was empty", nil)
+		os.Exit(1)
+	}
+}
+
+// Ensures that a variable isn't empty or only white space.
+func not_empty(variable *string) bool {
+	*variable = strings.TrimSpace(*variable)
+	if *variable == "" {
+		return false
+	} else {
+		return true
+	}
+}
+
+func log_error(message string, vopr_message_hash []byte) {
+	log_message("error: ", message, vopr_message_hash)
+}
+
+func log_debug(message string, vopr_message_hash []byte) {
+	if !debug_mode {
+		return
+	} else {
+		log_message("debug: ", message, vopr_message_hash)
+	}
+}
+
+func log_info(message string, vopr_message_hash []byte) {
+	log_message("info:  ", message, vopr_message_hash)
+}
+
+// Formats all the log messages and adds a timestamp to them.
+func log_message(log_level string, message string, vopr_message_hash []byte) {
+	// Gets the current time in UTC and rounds to the nearest second.
+	timestamp := time.Now().UTC().Round(time.Second).Format("2006-01-02 15:04:05.999999999")
+	if vopr_message_hash != nil {
+		// Only use the first 16 bytes of the message hash as the ID
+		fmt.Printf(timestamp+" "+log_level+"ID:%x %s\n", vopr_message_hash[0:16], message)
+	} else {
+		fmt.Printf(timestamp+" "+log_level+"%s\n", message)
+	}
+}
+
+func main() {
+	// Determine the mode in which to run the VOPR Hub
+	flag.BoolVar(&debug_mode, "debug", false, "runs with debugging logs enabled")
+	flag.Parse()
+
+	set_environment_variables()
+
+	// This channel ensures no more than MAX_CONCURRENT_CONNECTIONS are being handled at one time.
+	track_connections := make(chan bool, MAX_CONCURRENT_CONNECTIONS)
+
+	// The channel will receieve fixed-size byte arrays from a VOPR.
+	vopr_message_channel := make(chan vopr_message, MAX_QUEUING_MESSAGES)
+
+	// Create a worker Goroutine to call process on each item as it appears in the channel.
+	go worker(vopr_message_channel)
+
+	listener, error := net.Listen("tcp", vopr_hub_address)
+	if error != nil {
+		error_message := fmt.Sprintf("Could not listen on %s: %s", vopr_hub_address, error.Error())
+		log_error(error_message, nil)
+		os.Exit(1)
+	}
+	defer listener.Close()
+	log_info("Listening...", nil)
+
+	for {
+		connection, error := listener.Accept()
+		connection.SetReadDeadline(time.Now().Add(10 * time.Second))
+		if error != nil {
+			error_message := fmt.Sprintf(
+				"Unable to set read deadline on the connection: %s",
+				error.Error(),
+			)
+			log_error(error_message, nil)
+			panic(error.Error())
```

Comment:
Could this be a `connection.Close(); continue;` instead of a `panic()`? 

If a connection is Accept()ed, and is immediately closed right after, `SetReadDeadline()` could return `error != nil` and panic which would bring down the server.